gossipsub.go

```diff
@@ -0,0 +1,569 @@
+			gmap, ok = gs.fanout[topic]
+			if !ok {
+				// we don't have any, pick some
+				peers := gs.getPeers(topic, GossipSubD, func(peer.ID) bool { return true })
+
+				if len(peers) > 0 {
+					gmap = peerListToMap(peers)
+					gs.fanout[topic] = gmap
+				}
+			}
+		}
+
+		for p := range gmap {
+			tosend[p] = struct{}{}
+		}
+	}
+
+	out := rpcWithMessages(msg)
+	for pid := range tosend {
+		if pid == from || pid == peer.ID(msg.GetFrom()) {
+			continue
+		}
+
+		gs.sendRPC(pid, out)
+	}
+}
+
+func (gs *GossipSubRouter) Join(topic string) {
+	gmap, ok := gs.mesh[topic]
+	if ok {
+		return
+	}
+
+	gmap, ok = gs.fanout[topic]
+	if ok {
+		gs.mesh[topic] = gmap
+		delete(gs.fanout, topic)
+	} else {
+		peers := gs.getPeers(topic, GossipSubD, func(peer.ID) bool { return true })
+		gmap = peerListToMap(peers)
+		gs.mesh[topic] = gmap
+	}
+
+	for p := range gmap {
+		gs.sendGraft(p, topic)
+	}
+}
+
+func (gs *GossipSubRouter) Leave(topic string) {
+	gmap, ok := gs.mesh[topic]
+	if !ok {
+		return
+	}
+
+	delete(gs.mesh, topic)
+
+	for p := range gmap {
+		gs.sendPrune(p, topic)
+	}
+}
+
+func (gs *GossipSubRouter) sendGraft(p peer.ID, topic string) {
+	graft := []*pb.ControlGraft{&pb.ControlGraft{TopicID: &topic}}
+	out := rpcWithControl(nil, nil, nil, graft, nil)
+	gs.sendRPC(p, out)
+}
+
+func (gs *GossipSubRouter) sendPrune(p peer.ID, topic string) {
+	prune := []*pb.ControlPrune{&pb.ControlPrune{TopicID: &topic}}
+	out := rpcWithControl(nil, nil, nil, nil, prune)
+	gs.sendRPC(p, out)
+}
+
+func (gs *GossipSubRouter) sendRPC(p peer.ID, out *RPC) {
+	// piggyback cotrol message retries
+	ctl, ok := gs.control[p]
+	if ok {
+		gs.piggybackControl(p, out, ctl)
+		delete(gs.control, p)
+	}
+
+	// piggyback gossip
+	ihave, ok := gs.gossip[p]
+	if ok {
+		gs.piggybackGossip(p, out, ihave)
+		delete(gs.gossip, p)
+	}
+
+	mch, ok := gs.p.peers[p]
+	if !ok {
+		return
+	}
+
+	select {
+	case mch <- out:
+	default:
+		log.Infof("dropping message to peer %s: queue full", p)
+		// push control messages that need to be retried
+		ctl := out.GetControl()
+		if ctl != nil {
+			gs.pushControl(p, ctl)
+		}
+	}
+}
+
+func (gs *GossipSubRouter) heartbeatTimer() {
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			select {
+			case gs.p.eval <- gs.heartbeat:
+			case <-gs.p.ctx.Done():
+				return
+			}
+		case <-gs.p.ctx.Done():
+			return
+		}
+	}
+}
+
+func (gs *GossipSubRouter) heartbeat() {
+	// flush pending control message from retries and gossip
+	// that hasn't been piggybacked since the last heartbeat
+	gs.flush()
+
+	tograft := make(map[peer.ID][]string)
+	toprune := make(map[peer.ID][]string)
+
+	// maintain the mesh for topics we have joined
+	for topic, peers := range gs.mesh {
+
+		// do we have enough peers?
+		if len(peers) < GossipSubDlo {
+			ineed := GossipSubD - len(peers)
+			plst := gs.getPeers(topic, ineed, func(p peer.ID) bool {
+				_, ok := peers[p]
```

Comment:
this looks like we're filtering something, mind adding a comment on what we're filtering on?

Reply:
we are filtering the peers that are not already in our peer list; i can add a comment to that extent.